pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,771 @@
+	file, writer, err := getPacketFileAndWriter(pc.Name)
+	if err != nil {
+		return err
+	}
+	pcState.pcapngFile = file
+	pcState.pcapngWriter = writer
+	pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+	go func() {
+		captureErr := c.performCapture(ctx, pc, pcState, *device, srcIP, dstIp)
+		func() {
+			c.lock.Lock()
+			defer c.lock.Unlock()
+			c.numRunningCaptures -= 1
+			state := c.captures[pc.Name]
+			state.phase = packetCapturePhaseCompleted
+			state.err = captureErr
+		}()
+		c.addPacketCapture(pc)
+	}()
+	return nil
+}
+
+func (c *Controller) performCapture(
+	ctx context.Context,
+	pc *crdv1alpha1.PacketCapture,
+	captureState *packetCaptureState,
+	device string,
+	srcIP, dstIP net.IP,
+) error {
+	packets, err := c.captureInterface.Capture(ctx, device, srcIP, dstIP, pc.Spec.Packet)
+	if err != nil {
+		klog.ErrorS(err, "Failed to start capture")
+		return err
+	}
+
+	for {
+		select {
+		case packet := <-packets:
+			c.lock.Lock()
+			if captureState.isCaptureSucceed() {
+				c.lock.Unlock()
+				return nil
+			}
+			captureState.capturedPacketsNum++
+			c.lock.Unlock()
+			ci := gopacket.CaptureInfo{
+				Timestamp:     time.Now(),
+				CaptureLength: len(packet.Data()),
+				Length:        len(packet.Data()),
+			}
+			err = captureState.pcapngWriter.WritePacket(ci, packet.Data())
+			if err != nil {
+				return fmt.Errorf("couldn't write packets: %w", err)
+			}
+			klog.V(5).InfoS("Capture packets", "name", captureState.name, "count",
+				captureState.capturedPacketsNum, "len", ci.Length)
+
+			c.lock.Lock()
+			reachTarget := captureState.isCaptureSucceed()
+			c.lock.Unlock()
+			// use rate limiter to reduce the times we need to update status.
+			if reachTarget || captureState.updateRateLimiter.Allow() {
+				pc, err := c.packetCaptureLister.Get(captureState.name)
+				if err != nil {
+					return fmt.Errorf("get PacketCapture failed: %w", err)
+				}
+				// if reach the target. flush the file and upload it.
+				if reachTarget {
+					path := env.GetPodName() + ":" + nameToPath(pc.Name)
+					statusPath := path
+					if err = captureState.pcapngWriter.Flush(); err != nil {
+						return err
+					}
+					if pc.Spec.FileServer != nil {
+						err = c.uploadPackets(pc, captureState.pcapngFile)
+						klog.V(4).InfoS("Upload captured packets", "name", pc.Name, "path", path)
+						statusPath = fmt.Sprintf("%s/%s.pcapng", pc.Spec.FileServer.URL, pc.Name)
+					}
+					c.lock.Lock()
+					captureState.filePath = statusPath
+					c.lock.Unlock()
+					if err != nil {
+						return err
+					}
+					if err := captureState.pcapngFile.Close(); err != nil {
+						klog.ErrorS(err, "Close pcapng file error", "name", pc.Name, "path", path)
+					}
+				}
+				// report capture status.
+				c.addPacketCapture(pc)
+			}
+		case <-ctx.Done():
+			pc, err := c.packetCaptureLister.Get(captureState.name)
+			if err != nil {
+				return fmt.Errorf("get PacketCapture failed: %w", err)
+			}
```

Comment:
I don't see it necessary to query Lister once. When this happens, either the CR is deleted, or it times out, and it could just return `ctx.Err()`